### PR TITLE
Split and move `@num.Num` into `@math.Num` parts

### PR DESCRIPTION
**Pull request by fantix**
As mentioned in [#36](https://github.com/moonbitlang/core/pull/36#issuecomment-2027375604), this PR proposes breaking down the `Num` trait into finer pieces so that the implementations don't always have to fulfill the complete `Num` trait.

Refs #49

### CODE REVIEW

File: ref/ref.mbti [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
+alias @moonbitlang/core/math as @math
 
 // Values
 fn ref[T](T) -> Ref[T]
 
 // Types and methods
-fn Ref::decr[T : @num.Num](Ref[T], T) -> Unit
-fn Ref::incr[T : @num.Num](Ref[T], T) -> Unit
+fn Ref::decr[T : @math.FromInt + @math.Subtract](Ref[T], T) -> Unit
+fn Ref::incr[T : @math.FromInt + @math.Add](Ref[T], T) -> Unit
```
Comment (bobzhang) [inline ref/ref.mbti:10]: in general, you want incr to be really efficient,
e.g,
```
let  v = ref (0)
incr(v)
```
people would expect `v` stack allocated

Comment (fantix) [reply ref/ref.mbti]: Now rebased to include #335

**Comment by bobzhang** [comment]
feel free to reopen it if you have continued interest

**Comment by fantix** [comment]
Sure, though, this has now been moved to X, right?

**Pull request closed without merging**